modules/database/sqlite/wp-includes/sqlite/class-perflab-sqlite-db.php

```diff
@@ -0,0 +1,329 @@
+	}
+
+	/**
+	 * Method to flush cached data.
+	 *
+	 * This overrides wpdb::flush(). This is not necessarily overridden, because
+	 * $result will never be resource.
+	 *
+	 * @see wpdb::flush
+	 */
+	public function flush() {
+		$this->last_result   = array();
+		$this->col_info      = null;
+		$this->last_query    = null;
+		$this->rows_affected = 0;
+		$this->num_rows      = 0;
+		$this->last_error    = '';
+		$this->result        = null;
+	}
+
+	/**
+	 * Method to do the database connection.
+	 *
+	 * This overrides wpdb::db_connect() to avoid using MySQL function.
+	 *
+	 * @see wpdb::db_connect()
+	 *
+	 * @param bool $allow_bail Not used.
+	 */
+	public function db_connect( $allow_bail = true ) {
+		$this->init_charset();
+		$this->dbh   = new Perflab_SQLite_PDO_Engine();
+		$this->ready = true;
+	}
+
+	/**
+	 * Method to dummy out wpdb::check_connection()
+	 *
+	 * @param bool $allow_bail Not used.
+	 *
+	 * @return bool
+	 */
+	public function check_connection( $allow_bail = true ) {
+		return true;
+	}
+
+	/**
+	 * Method to execute the query.
+	 *
+	 * This overrides wpdb::query(). In fact, this method does all the database
+	 * access jobs.
+	 *
+	 * @see wpdb::query()
+	 *
+	 * @param string $query Database query.
+	 *
+	 * @return int|false Number of rows affected/selected or false on error
+	 */
+	public function query( $query ) {
+		if ( ! $this->ready ) {
+			return false;
+		}
+
+		$query = apply_filters( 'query', $query );
+
+		$this->flush();
+
+		$this->func_call  = "\$db->query(\"$query\")";
+		$this->last_query = $query;
+
+		if ( defined( 'SAVEQUERIES' ) && SAVEQUERIES ) {
+			$this->timer_start();
+		}
+
+		$this->result = $this->dbh->query( $query );
+		$this->num_queries++;
+
+		if ( defined( 'SAVEQUERIES' ) && SAVEQUERIES ) {
+			$this->queries[] = array( $query, $this->timer_stop(), $this->get_caller() );
+		}
+
+		$this->last_error = $this->dbh->get_error_message();
+		if ( $this->last_error && ( ! defined( 'WP_INSTALLING' ) || ! WP_INSTALLING ) ) {
+			$this->print_error( $this->last_error );
+			return false;
+		}
+
+		if ( preg_match( '/^\\s*(create|alter|truncate|drop|optimize)\\s*/i', $query ) ) {
+			return $this->dbh->get_return_value();
+		}
+
+		if ( preg_match( '/^\\s*(insert|delete|update|replace)\s/i', $query ) ) {
+			$this->rows_affected = $this->dbh->get_affected_rows();
+			if ( preg_match( '/^\s*(insert|replace)\s/i', $query ) ) {
+				$this->insert_id = $this->dbh->get_insert_id();
+			}
+			return $this->rows_affected;
+		}
+
+		$this->last_result = $this->dbh->get_query_results();
+		$this->num_rows    = $this->dbh->get_num_rows();
+		return $this->num_rows;
+	}
+
+	/**
+	 * Method to set the class variable $col_info.
+	 *
+	 * This overrides wpdb::load_col_info(), which uses a mysql function.
+	 *
+	 * @see    wpdb::load_col_info()
+	 * @access protected
+	 */
+	protected function load_col_info() {
+		if ( $this->col_info ) {
+			return;
+		}
+		$this->col_info = $this->dbh->get_columns();
+	}
+
```

Comment:
Is there a solid way for plugin authors to know this instance is on SQLite, rather than MariaDB / MySQL ? I'm thinking of backup plugins and the like.  Compatibility may be problematical for some.

Reply:
They can check for the value of the `DATABASE_TYPE` constant.
In this implementation I'm hardcoding it to `define( 'DATABASE_TYPE', 'sqlite' );` because if the module is active they obviously want to use an SQLite database, but when backported to Core this will be something defined in `wp-settings.php`.
For the purposes of this module, constants are defined in the `modules/database/sqlite/constants.php` file. 👍 

As a sidenote, backup plugins etc will be thrilled with SQLIte... all they have to do is backup the database file itself without making any database queries 😄